csrc/ops/utils.cpp

```diff
@@ -299,6 +338,26 @@ IterDomain* newOutputIterDomain(const std::vector<IterDomain*>& ids) {
 #pragma GCC diagnostic pop
 #endif
 
+std::vector<IterDomain*> newOutputDomain(const std::vector<std::vector<IterDomain*>>& input_ids) {
+  NVF_CHECK(
+      !input_ids.empty(),
+      "Tried to create new output Tensorview but received empty list.");
+
+  std::vector<IterDomain*> out_domain(input_ids.front().size(), nullptr);
+
+  for (const auto dim_i : c10::irange(out_domain.size())) {
+    std::vector<IterDomain*> ids_i;
+    ids_i.reserve(input_ids.size());
+    for (auto ids : input_ids) {
+      if (ids[dim_i] != nullptr){
+        ids_i.emplace_back(ids[dim_i]);
```

Comment:
This is what I meant by
> mapping_b will take precedence over mapping_bias.

`mapping_b` and `mapping_bias` would both have a mapping here for `N` dimension. So we are always taking `mapping_b`'s id for that dimension.

`torch.linear` doesn't support bias dominating the out_feature while `weight` has size-1 there, but that's just an implementation detail.

```
>>> x = torch.randn(2, 3)
>>> w = torch.randn(1, 3)
>>> b = torch.randn(5)
>>> torch.nn.functional.linear(x, w, b)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
RuntimeError: The expanded size of the tensor (1) must match the existing size (5) at non-singleton dimension 1.  Target sizes: [2, 1].  Tensor sizes: [5]
```

Reply:
I think the example I posted should *in theory* work for nvfuser scheduled kernel. But I don't think the code we have is working properly. I think your output iter domain will be created as
`x->axis(0), x->axis(1), w->axis(0)`
but it should have been
`x->axis(0), x->axis(1), b->axis(0)`

Bus since this is a case where pytorch asserts, maybe it doesn't matter. Having said that, we should at least throw an error in this case.

Reply:
> This is what I meant by
> 
> > mapping_b will take precedence over mapping_bias.
> 
> `mapping_b` and `mapping_bias` would both have a mapping here for `N` dimension. So we are always taking `mapping_b`'s id for that dimension.
> 
> `torch.linear` doesn't support bias dominating the out_feature while `weight` has size-1 there, but that's just an implementation detail.
> 
> ```
> >>> x = torch.randn(2, 3)
> >>> w = torch.randn(1, 3)
> >>> b = torch.randn(5)
> >>> torch.nn.functional.linear(x, w, b)
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
> RuntimeError: The expanded size of the tensor (1) must match the existing size (5) at non-singleton dimension 1.  Target sizes: [2, 1].  Tensor sizes: [5]
> ```

Yes, that is correct. Bias can be broadcast but if the weight is broadcast in N dimension, then bias can only be broadcast.
I took both the IDs to create the ID for the `out_features` since if we have a non-broadcast N dimension and weight is symbolic but bias is concrete tensor, then the extent for the N dimension in the output can be resolved.

Reply:
> I think the example I posted should _in theory_ work for nvfuser scheduled kernel. But I don't think the code we have is working properly. I think your output iter domain will be created as `x->axis(0), x->axis(1), w->axis(0)` but it should have been `x->axis(0), x->axis(1), b->axis(0)`
> 
> Bus since this is a case where pytorch asserts, maybe it doesn't matter. Having said that, we should at least throw an error in this case.

Right, we should assert on both being broadcast if weight has N=1

Reply:
Ha, sorry I was not reading this correctly. I didn't notice this is taking all the IDs. sorry about that.

:face_with_head_bandage: